### ISSUE TITLE
There is no become_method for Windows/WINRM/Powershell

### ISSUE THREAD

**Issue by dagwieers** [ansible/ansible#16202]
##### ISSUE TYPE

- Bug Report
##### ANSIBLE VERSION

v2.1
##### CONFIGURATION

``` ini
[defaults]
hostfile = hosts
transport = winrm
become_method = runas
remote_user = Administrator

[winrm]
server_cert_validation = ignore
```
##### OS / ENVIRONMENT

Windows 10
##### SUMMARY

If you have a user with no password (autologon) and you want to run something as this user (through the Administrator account), ansible uses the Administrator password automatically as the user password (because we pressed ENTER on the prompt, so it assumes the same password).
##### STEPS TO REPRODUCE

- Create a password-less local Windows account
- Try to run a playbook as Administrator and attempt to run something as this local user
##### EXPECTED RESULTS

```
[dag@moria crombeen]$ ansible-playbook -kK -l computer66 bginfo.yml 
SSH password: 
SUDO password[defaults to SSH password]:

PLAY [all] *********************************************************************

TASK [Run bginfo as user] ******************************************************
ok: [computer65]

PLAY RECAP *********************************************************************
computer65                 : ok=1    changed=0    unreachable=0    failed=0   
```
##### ACTUAL RESULTS

```
[dag@moria crombeen]$ ansible-playbook -k -l computer65 bginfo.yml 
SSH password: <Typed password>
SUDO password[defaults to SSH password]: <Pressed ENTER for empty password>

PLAY [all] *********************************************************************

TASK [Run bginfo as user] ******************************************************
fatal: [computer65]: FAILED! => {"failed": true, "msg": "ssl: the specified credentials were rejected by the server"}

NO MORE HOSTS LEFT *************************************************************
        to retry, use: --limit @bginfo.retry

PLAY RECAP *********************************************************************
computer65                 : ok=0    changed=0    unreachable=0    failed=1   
```

**Comment by bcoca** [comment]
no become methods are currently supported on windows (also it would be runas , not doas)

**Comment by dagwieers** [comment]
That's quite  ~~funny~~ dramatic, because:
1. None of this is documented (ie. there is no description of each method, only a list, so I assumed...)
2. The become method didn't complain or fail, it appeared to be working (but silently failing)
3. That's quite an important limitation for Windows (which is not documented either)

The dramatic part being I spend quite some time/effort on this project to now conclude that part of what I wanted isn't even possible. (Managing Windows desktops without runas is impossible)

Let's rename the issue then....

**Comment by bcoca** [comment]
Agreed, we are sorely missing in docs in this area, windows currently does not support ANY method of privilege escalation, we had plans to enable 'runas' which is the windows specific one (doas is openbsd only afaik).

I was under the impression that to use winrm users just assigned admin privileges to the user connecting, which removes most cases of needing privilege escalation.

**Comment by dagwieers** [comment]
Which means you cannot manage unprivileged users (in our case unprivileged users with no password) in Ansible. Given that we manage Windows desktops in an elementary school, almost everything we would like to manage (except software install) has to be performed as the unprivileged user...

**Comment by nitzmahone** [comment]
This _should_ have thrown an error, as we don't support any become methods over winrm yet (I've prototyped a couple different things, but nothing ready for prime-time yet). Looks like we're not calling the base Connection's `_connect()` method (where the check occurs). Fixed in https://github.com/ansible/ansible/commit/445a88d3e8af0a11b63ecb4958cb16d2a27e4716 and cherry-picked to stable-2.1.

**Comment by nitzmahone** [comment]
Ack, GH didn't refresh for me.

**Comment by nitzmahone** [comment]
@dagwieers I got it working with a passwordless, unprivileged user... Had to slightly modify the required password check in pywinrm (to check for None instead of `not self.password`) and add the right privs to the unprivileged user with `winrm configSDDL default`, but you would've had to do that part anyway (winrm restriction, not ours). Would that take care of your needs?

**Comment by nitzmahone** [comment]
Bigger picture question: why a passwordless user instead of a passworded one with autologin?

**Comment by dagwieers** [comment]
@nitzmahone How did you make it work without a `runas` implementation ? My initial report was flawed since I was using `doas` and it didn't work... :-/

In an elementary school we have a passwordless user with autologin for the students. I guess it is passwordless because if you would log out, a password is needed to log in again. To be honest, I think the autologin is based on the fact that the user has no password, because I didn't specify autologin anywhere...

PS I'd like to investigate to reset the user on logout/reboot to a default state. Not looked at that yet though.

**Comment by nitzmahone** [comment]
In addition to the minor password validation change in pywinrm (already committed to [pywinrm master](https://github.com/diyan/pywinrm/tree/master)), I had to disable `Accounts: Limit local account use of blank passwords to console logon only` under Security Options in group policy.

**Comment by nitzmahone** [comment]
A runas (ish) become method could make this _slightly_ more secure, since you wouldn't need to allow remote access to the passwordless user- an administrative remote login would be required to modify the unprivileged user remotely. Grade-school kids seem unlikely to exploit this over just dorking with it via local access (though I probably would've tried- I was a li'l bastard about stuff like that). 

The "reset" thing could be tricky- bums me out that MS keeps killing the true "temporary guest account" concept, because that's _exactly_ what you need. I'm not sure Ansible is the best choice for that part- a logoff script that creates and rotates to a new autologin user might be better. You could do it with something like a Tower provisioning callback, but there'd be inherent races involved, even if you invoked it synchronously via a logoff script.

**Comment by nitzmahone** [comment]
(PS, now that we've fixed the "become doesn't throw an error thing", I'm leaving this open as a feature request)

**Comment by dagwieers** [comment]
@nitzmahone Right, I don't want remote access to passwordless users. (I didn't even know that was possible !) And indeed, I wasn't going to rely on Ansible for a "temporary guest account", at least not for the clearing/rotation.

I do think `runas` is an important missing piece in the Windows implementation and essential for bringing it up to par with other OSes. (Maybe high on Red Hat's Windows wishlist ? ;-))

**Comment by nitzmahone** [comment]
I agree that it's a necessity for platform parity, but strangely, it hasn't come up very often thus far- I think you're the second direct request I've gotten for it. People ask about it early on when gathering information, but once they start actually using Ansible, they end up doing most Windows remote mgmt directly with privileged users. Perhaps it's more about the modules we ship right now (many of them are useless w/o admin privs), and as people start doing deeper management of individual user profiles, it'll come up more.

As I said, I've prototyped a couple different approaches (I'm not counting actual `runas.exe`\- that's a non-starter for several reasons)- there are various issues that currently prevent any of them from being a slam-dunk or one-size-fits-all. Given how busy 2.2 is already, it's probably not going to make it into the 2.2 roadmap, but if there's enough other clamor for it, could be in for 2.3.

**Comment by tim-seoss** [comment]
Any details on what's been prototyped for this?  It would be good to know which approaches are thought not to be worth pursuing, and why (having been banging my head against this for a couple of days before seeing this bug).

**Comment by tim-seoss** [comment]
FWIW, I wanted to create a network drive mapping as an unprivileged user, the user does have a password.  The machines are standalone.

I tried using a powershell script like this:

$username = "xyz"
$password = "abc"
$credentials = New-Object System.Management.Automation.PSCredential -ArgumentList @($username,(ConvertTo-SecureString -String $password -AsPlainText -Force))
Start-Process powershell.exe -Credential ($credentials) -NoNewWindow -WorkingDirectory C:\Windows\Temp -LoadUserProfile -ArgumentList "-ExecutionPolicy ByPass -File .\map-net-drive.ps1" -Wait

This script works when:

. Executed as the normal ansible user (which is in the administrators group) from a command prompt which was started with "Run as Administrator"
. Executed from an ssh session using win32-openssh (installed via chocolately package, with the manual fixes detailed in https://github.com/PowerShell/Win32-OpenSSH/issues/227 ) if logged in as the unprivileged user (using password auth for ssh - I didn't try ssh key auth)

It doesn't work when:

. Executed as the normal ansible user (in admin group) on the console, using a cmd prompt which was not started with "Run as Administrator".
. Executed as the normal ansible user (in admin group) via winrm.

**Comment by ccoager** [comment]
:+1:
Would like to see this implemented as well. Not having this severely limits functionality on Windows where UAC is mandatory.

Does anyone have a work around in the mean time?

**Comment by jhawkesworth** [comment]
@tim-seoss - I think the issue you are describing is a 'second hop' issue.  There's one authentication hop from ansible to your target windows machine, and a second hop from the target machine to the network share.  The only way round this right now is using kerberos with auth delegation enabled.

@ccoager - I believe UAC doesn't apply for privileged accounts (e.g. local administrators) over winrm, so simplest way to get around is to connect using a suitably privileged user.  You can temporarily disable UAC for other users by modifying the registry if needed too.

**Comment by dagwieers** [comment]
@jhawkesworth But most modules don't work for modifying user-accounts (i.e. modifying registry settings for specific users, mapping disks for users with credentials, etc.). Especially if you're managing desktops the lack of become_method for Windows is severely limiting what we can do.

**Comment by jhawkesworth** [comment]
@dagwieers I'm not saying this isn't needed, just that there are ways around some of this stuff.
If you are happy passing credentials as module parameters and using powershell, you can do some things that require other user credentials.  From memory win_package does this, although perhaps not the best module to look at as its mostly generated from a dsc resource.

**Comment by ccoager** [comment]
@dagwieers
UAC can apply to administrators. There are separate policies for regular users vs. administrators. I haven't tried the registry setting yet, does that require a reboot?

**Comment by nitzmahone** [comment]
fixed by #21510

**Comment by dagwieers** [comment]
\o/
